internal/grpc/services/publicstorageprovider/publicstorageprovider.go

```diff
@@ -111,11 +128,176 @@ func (s *service) UnsetArbitraryMetadata(ctx context.Context, req *provider.Unse
-	return nil, gstatus.Errorf(codes.Unimplemented, "method not implemented")
+	statReq := &provider.StatRequest{Ref: req.Ref}
+	statRes, err := s.Stat(ctx, statReq)
+	if err != nil {
+		return &provider.InitiateFileDownloadResponse{
+			Status: status.NewInternal(ctx, err, "gateway: error stating ref:"+req.Ref.String()),
+		}, nil
+	}
+	if statRes.Status.Code != rpc.Code_CODE_OK {
+		if statRes.Status.Code == rpc.Code_CODE_NOT_FOUND {
+			return &provider.InitiateFileDownloadResponse{
+				Status: status.NewNotFound(ctx, "gateway: file not found"),
+			}, nil
+		}
+		err := status.NewErrorFromCode(statRes.Status.Code, "gateway")
+		return &provider.InitiateFileDownloadResponse{
+			Status: status.NewInternal(ctx, err, "gateway: error stating ref"),
+		}, nil
+	}
+
+	req.Opaque = statRes.Info.Opaque
+	return s.initiateFileDownload(ctx, req)
+}
+
+// PathTranslator encapsulates behavior that requires translation between external paths into internal paths. Internal
+// path depend on the storage provider in use, hence a translation is needed. In the case of Public Links, the translation
+// goes beyond a simple path conversion but instead, the token needs to be "expanded" to an internal path. Essentially
+// transforming:
+// "YzUTlrKrpswo/foldera/folderb/file.txt"
+// into:
+// "shared-folder-path/internal-folder/foldera/folderb/file.txt".
+type PathTranslator struct {
+	dir   string
+	base  string
+	token string
+}
+
+// NewPathTranslator creates a new PathTranslator.
+func NewPathTranslator(p string) *PathTranslator {
+	return &PathTranslator{
+		dir:   filepath.Dir(p),
+		base:  filepath.Base(p),
+		token: strings.Split(p, "/")[2],
+	}
+}
+
+// Both, t.dir and tokenPath paths need to be merged:
+// tokenPath   = /oc/einstein/public-links
+// t.dir       = /public/ausGxuUePCOi/foldera/folderb/
+// res         = /public-links/foldera/folderb/
+// this `res` will get then expanded taking into account the authenticated user and the storage:
+// end         = /einstein/files/public-links/foldera/folderb/
+
+// TODO how would this behave with a storage other than owncloud?
+// the current OC namespace looks like "/oc/einstein/public-links"
+// and operations on it would be hardcoded string operations bound to
+// the storage path. Either we need clearly defined paths from each storage,
+// or else this would remain black magic.
+func (s *service) initiateFileDownload(ctx context.Context, req *provider.InitiateFileDownloadRequest) (*provider.InitiateFileDownloadResponse, error) {
+	t := NewPathTranslator(req.Ref.GetPath())
+	tokenPath, err := s.pathFromToken(ctx, t.token)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = pool.GetPublicShareProviderClient(s.conf.DriverAddr)
+	if err != nil {
+		err = errors.Wrap(err, "gateway: error getting a storage provider client")
+		return nil, err
+	}
+
+	if isOCStorage(tokenPath) {
+		base := strings.Join(strings.Split(tokenPath, "/")[3:], "/")
```

Comment:
are paths normalized ? could there be leaks in case of double slashes ?
does Golang have a library for properly handling paths and deduplicating separators ?

Reply:
the `path` package